expression/src/display.rs

```diff
@@ -0,0 +1,151 @@
+            _ => None,
+        }
+    }
+}
+
+impl<T: Display, R: Display> Display for AlgebraicExpression<T, R> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            AlgebraicExpression::Reference(reference) => write!(f, "{reference}"),
+            AlgebraicExpression::Number(value) => write!(f, "{value}"),
+            AlgebraicExpression::BinaryOperation(o) => {
+                write!(f, "{o}")
+            }
+            AlgebraicExpression::UnaryOperation(o) => write!(f, "{o}"),
+        }
+    }
+}
+
+impl<T: Display, R: Display> Display for AlgebraicBinaryOperation<T, R> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let op_precedence = self.op.precedence().unwrap();
+        let use_left_parentheses = match self.left.precedence() {
+            Some(left_precedence) => left_precedence > op_precedence,
+            None => false,
+        };
+
+        let use_right_parentheses = match self.right.precedence() {
+            Some(right_precedence) => right_precedence >= op_precedence,
+            None => false,
+        };
+
+        let left_string = if use_left_parentheses {
+            format!("({})", self.left)
+        } else {
+            format!("{}", self.left)
+        };
+        let right_string = if use_right_parentheses {
+            format!("({})", self.right)
+        } else {
+            format!("{}", self.right)
+        };
+
+        write!(f, "{left_string} {} {right_string}", self.op)
+    }
+}
+
+impl<T: Display, R: Display> Display for AlgebraicUnaryOperation<T, R> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let exp_string = match (self.op.precedence(), self.expr.precedence()) {
+            (Some(precedence), Some(inner_precedence)) if precedence < inner_precedence => {
+                format!("({})", self.expr)
+            }
+            _ => {
+                format!("{}", self.expr)
+            }
+        };
+
+        write!(f, "{}{exp_string}", self.op)
+    }
+}
+
+impl Display for AlgebraicUnaryOperator {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            AlgebraicUnaryOperator::Minus => write!(f, "-"),
+        }
+    }
+}
+
+impl Display for AlgebraicBinaryOperator {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            AlgebraicBinaryOperator::Add => write!(f, "+"),
+            AlgebraicBinaryOperator::Sub => write!(f, "-"),
+            AlgebraicBinaryOperator::Mul => write!(f, "*"),
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use powdr_number::GoldilocksField;
+    use pretty_assertions::assert_eq;
+    use test_log::test;
+
+    use super::AlgebraicExpression;
+
+    fn test_display(expr: AlgebraicExpression<GoldilocksField, &str>, expected: &str) {
+        assert_eq!(expr.to_string(), expected);
+    }
+
+    #[test]
+    fn binary_op() {
```

Comment:
Ported from [here](https://github.com/powdr-labs/powdr/blob/53ec1878439a5a0c15304eaafdf79729a0089d21/ast/src/analyzed/display.rs#L682). I believe these are the only relevant tests from this module. The code has been adjusted from the ast Expression, because the removal of the next operator and `pow` operation simplified stuff. It should be carefully reviewed.

Reply:
Could use some tests with other terminals, but not necessarily.